test/integration/ocsp_test.go

```diff
@@ -0,0 +1,192 @@
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("adding http-01 response: status %d", resp.StatusCode)
+	}
+	resp.Body.Close()
+	return nil
+}
+
+type issuanceResult struct {
+	acme.Order
+	certs []*x509.Certificate
+}
+
+func authAndIssue(domains []string) (*issuanceResult, error) {
+	c, err := makeClient()
+	if err != nil {
+		return nil, err
+	}
+	var ids []acme.Identifier
+	for _, domain := range domains {
+		ids = append(ids, acme.Identifier{Type: "dns", Value: domain})
+	}
+	order, err := c.Client.NewOrder(c.Account, ids)
+	if err != nil {
+		return nil, fmt.Errorf("making order: %s", err)
+	}
+
+	for _, authUrl := range order.Authorizations {
+		auth, err := c.Client.FetchAuthorization(c.Account, authUrl)
+		if err != nil {
+			return nil, fmt.Errorf("fetching authorization at %s: %s", authUrl, err)
+		}
+
+		chal, ok := auth.ChallengeMap[acme.ChallengeTypeHTTP01]
+		if !ok {
+			return nil, fmt.Errorf("no HTTP challenge at %s", authUrl)
+		}
+
+		err = addHTTP01Response(chal.Token, chal.KeyAuthorization)
+		if err != nil {
+			return nil, fmt.Errorf("adding HTTP-01 response: %s", err)
+		}
+		chal, err = c.Client.UpdateChallenge(c.Account, chal)
+		if err != nil {
+			return nil, fmt.Errorf("updating challenge: %s", err)
+		}
+	}
+
+	certKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("generating certificate key: %s", err)
+	}
+
+	csrDer, err := x509.CreateCertificateRequest(rand.Reader, &x509.CertificateRequest{
+		SignatureAlgorithm: x509.ECDSAWithSHA256,
+		PublicKeyAlgorithm: x509.ECDSA,
+		PublicKey:          certKey.Public(),
+		Subject:            pkix.Name{CommonName: domains[0]},
+		DNSNames:           domains,
+	}, certKey)
+	if err != nil {
+		return nil, fmt.Errorf("making csr: %s", err)
+	}
+	csr, err := x509.ParseCertificateRequest(csrDer)
+	if err != nil {
+		return nil, fmt.Errorf("parsing csr: %s", err)
+	}
+
+	order, err = c.Client.FinalizeOrder(c.Account, order, csr)
+	if err != nil {
+		return nil, fmt.Errorf("finalizing order: %s", err)
+	}
+	certs, err := c.Client.FetchCertificates(c.Account, order.Certificate)
+	if err != nil {
+		return nil, fmt.Errorf("fetching certificates: %s", err)
+	}
+	return &issuanceResult{order, certs}, nil
+}
+
+func TestPrecertificateOCSP(t *testing.T) {
+	domain := random_domain()
+	for _, port := range []int{4500, 4501, 4510, 4511} {
+		url := fmt.Sprintf("http://boulder:%d/add-reject-host", port)
+		body := []byte(fmt.Sprintf(`{"host": "%s"}`, domain))
+		resp, err := http.Post(url, "", bytes.NewBuffer(body))
+		if err != nil {
+			t.Fatalf("adding reject host: %s", err)
+		}
+		if resp.StatusCode != http.StatusOK {
+			t.Fatalf("adding reject host: %d", resp.StatusCode)
+		}
+		resp.Body.Close()
+	}
+
+	os.Setenv("DIRECTORY", "http://boulder:4001/directory")
+	_, err := authAndIssue([]string{domain})
+	if err != nil {
+		if strings.Contains(err.Error(), "urn:ietf:params:acme:error:serverInternal") &&
```

Comment:
Does the lib not export a problem type that can be used to check the detail and type without matching on the error msg string?

Reply:
There is a `Problem` type but it's being flattened by the `fmt.Errorf` calls in `authAndIssue`. Proposal: rather than undo the `fmt.Errorf` wrapping (which I think is useful), I'd like to keep the code as-is and then look at whether the new error wrapping stuff in Go 1.13 lets us have our cake and eat it too.

Reply:
WFM :+1: